File: backend/src/api/carts.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.api.deps import get_current_user, get_db
from src.persistence.models import User
from src.schemas.carts import (
    AddToCartRequest,
    CartResponse,
    CartTotalResponse,
    UpdateCartItemRequest,
)
from src.services.cart_service import CartService

router = APIRouter(prefix="/carts", tags=["carts"])


@router.get("", response_model=CartResponse)
async def get_cart(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CartResponse:
    """Get current user's shopping cart."""
    service = CartService(db)
    cart = await service.get_user_cart(current_user.id)
    return CartResponse.model_validate(cart)


@router.post("/items", response_model=CartResponse)
async def add_to_cart(
    request: AddToCartRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CartResponse:
    """Add item to cart."""
    service = CartService(db)
    try:
        await service.add_to_cart(current_user.id, request.product_id, request.quantity)
        cart = await service.get_user_cart(current_user.id)
        return CartResponse.model_validate(cart)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
        )


@router.delete("/items/{item_id}", response_model=CartResponse)
async def remove_from_cart(
    item_id: uuid.UUID,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CartResponse:
    """Remove item from cart."""
    service = CartService(db)
    success = await service.remove_from_cart(current_user.id, item_id)
    if not success:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail="Cart item not found"
        )
    cart = await service.get_user_cart(current_user.id)
    return CartResponse.model_validate(cart)


@router.patch("/items/{item_id}", response_model=CartResponse)
async def update_cart_item(
    item_id: uuid.UUID,
    request: UpdateCartItemRequest,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CartResponse:
    """Update cart item quantity. Use quantity=0 to remove."""
    service = CartService(db)
    try:
        await service.update_cart_item(current_user.id, item_id, request.quantity)
        cart = await service.get_user_cart(current_user.id)
        return CartResponse.model_validate(cart)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST, detail=str(e)
        )


@router.delete("", response_model=CartResponse)
async def clear_cart(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CartResponse:
    """Clear all items from cart."""
    service = CartService(db)
    await service.clear_cart(current_user.id)
    cart = await service.get_user_cart(current_user.id)
    return CartResponse.model_validate(cart)


@router.get("/total", response_model=CartTotalResponse)
async def get_cart_total(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db),
) -> CartTotalResponse:
    """Get total price for user's cart."""
    service = CartService(db)
    total = await service.get_cart_total(current_user.id)
    return CartTotalResponse(total=total)

File: frontend/src/services/carts.ts
import apiClient from "./apiClient";
import { tokenStorage } from "./tokenStorage";
import { productAPI } from "./products";

export interface CartItem {
  id: string;
  product_id: string;
  quantity: number;
  price_at_add: number;
  created_at: string;
}

export interface CartResponse {
  id: string;
  user_id: string;
  items: CartItem[];
  created_at: string;
  updated_at: string;
}

export interface CartTotalResponse {
  total: number;
}

const GUEST_CART_KEY = "petty_cart";

const readGuestCart = (): CartItem[] => {
  try {
    const raw = localStorage.getItem(GUEST_CART_KEY);
    if (!raw) {
      return [];
    }

    const parsed = JSON.parse(raw) as CartItem[];
    return Array.isArray(parsed) ? parsed : [];
  } catch {
    return [];
  }
};

const writeGuestCart = (items: CartItem[]) => {
  localStorage.setItem(GUEST_CART_KEY, JSON.stringify(items));
};

const emitCartUpdated = () => {
  window.dispatchEvent(new Event("cart-updated"));
};

const isAuthenticated = () => Boolean(tokenStorage.getToken());

const getGuestCartResponse = (): CartResponse => {
  const items = readGuestCart();
  const now = new Date().toISOString();
  return {
    id: "guest-cart",
    user_id: "guest",
    items,
    created_at: now,
    updated_at: now,
  };
};

export const cartAPI = {
  /**
   * Get current user's shopping cart
   */
  getCart: async (): Promise<CartResponse> => {
    if (!isAuthenticated()) {
      return getGuestCartResponse();
    }
    const response = await apiClient.get("/carts");
    return response.data;
  },

  /**
   * Add item to cart
   */
  addToCart: async (
    productId: string,
    quantity: number
  ): Promise<CartResponse> => {
    if (!isAuthenticated()) {
      const product = await productAPI.getById(productId);
      const items = readGuestCart();
      const existingIndex = items.findIndex((item) => item.product_id === productId);

      if (existingIndex >= 0) {
        items[existingIndex] = {
          ...items[existingIndex],
          quantity: items[existingIndex].quantity + quantity,
        };
      } else {
        items.push({
          id: `guest-${productId}`,
          product_id: productId,
          quantity,
          price_at_add: product.price,
          created_at: new Date().toISOString(),
        });
      }

      writeGuestCart(items);
      emitCartUpdated();
      return getGuestCartResponse();
    }

    const response = await apiClient.post("/carts/items", {
      product_id: productId,
      quantity,
    });
    emitCartUpdated();
    return response.data;
  },

  /**
   * Remove item from cart
   */
  removeFromCart: async (itemId: string): Promise<CartResponse> => {
    if (!isAuthenticated()) {
      const items = readGuestCart().filter((item) => item.id !== itemId);
      writeGuestCart(items);
      emitCartUpdated();
      return getGuestCartResponse();
    }

    const response = await apiClient.delete(`/carts/items/${itemId}`);
    emitCartUpdated();
    return response.data;
  },

  /**
   * Update item quantity in cart (use quantity=0 to remove)
   */
  updateCartItem: async (
    itemId: string,
    quantity: number
  ): Promise<CartResponse> => {
    if (!isAuthenticated()) {
      const items = readGuestCart();
      const nextItems = quantity <= 0
        ? items.filter((item) => item.id !== itemId)
        : items.map((item) => (item.id === itemId ? { ...item, quantity } : item));
      writeGuestCart(nextItems);
      emitCartUpdated();
      return getGuestCartResponse();
    }

    const response = await apiClient.patch(`/carts/items/${itemId}`, {
      quantity,
    });
    emitCartUpdated();
    return response.data;
  },

  /**
   * Clear all items from cart
   */
  clearCart: async (): Promise<CartResponse> => {
    if (!isAuthenticated()) {
      writeGuestCart([]);
      emitCartUpdated();
      return getGuestCartResponse();
    }

    const response = await apiClient.delete("/carts");
    emitCartUpdated();
    return response.data;
  },

  /**
   * Get total price for cart
   */
  getCartTotal: async (): Promise<CartTotalResponse> => {
    const response = await apiClient.get("/carts/total");
    return response.data;
  },
};
